File: problems/go/240522-rotation-point/code.go
package main

import (
	"fmt"
)

func main() {
	words1 := []string{"grape", "orange", "plum", "radish", "apple"}
	fmt.Println("Rotation point: ", findRotationPoint(words1))

	words2 := []string{"cape", "cake"}
	fmt.Println("Rotation point: ", findRotationPoint(words2))
}

func findRotationPoint(words []string) int {
	floorIndex := 0
	ceilingIndex := len(words) - 1

	for floorIndex < ceilingIndex {
		guessIndex := floorIndex + (ceilingIndex-floorIndex)/2

		if words[guessIndex] < words[0] {
			ceilingIndex = guessIndex
		} else {
			floorIndex = guessIndex
		}

		if floorIndex+1 == ceilingIndex {
			break
		}
	}

	return ceilingIndex
}

File: problems/javascript/24052022-rotation-point/code.js
function findRotationPoint(words) {
  // Find the rotation point in the vector
  let floorIndex = 0
  let ceilingIndex = words.length - 1

  while (floorIndex < ceilingIndex) {
    const halfDistance = Math.floor((ceilingIndex - floorIndex) / 2)
    const guessIndex = floorIndex + halfDistance

    if (words[guessIndex] < words[0]) {
      ceilingIndex = guessIndex
    } else {
      floorIndex = guessIndex
    }

    if (floorIndex + 1 === ceilingIndex) break
  }

  return ceilingIndex
}

// Tests

let desc = 'small array'
let actual = findRotationPoint(['cape', 'cake'])
let expected = 1
assertEquals(actual, expected, desc)

desc = 'medium array'
actual = findRotationPoint(['grape', 'orange', 'plum', 'radish', 'apple'])
expected = 4
assertEquals(actual, expected, desc)

desc = 'large array'
actual = findRotationPoint([
  'ptolemaic',
  'retrograde',
  'supplant',
  'undulate',
  'xenoepist',
  'asymptote',
  'babka',
  'banoffee',
  'engender',
  'karpatka',
  'othellolagkage',
])
expected = 5
assertEquals(actual, expected, desc)

function assertEquals(a, b, desc) {
  if (a === b) {
    console.log(`${desc} ... PASS`)
  } else {
    console.log(`${desc} ... FAIL: ${a} != ${b}`)
  }
}
